### PR TITLE
Emit build operations around container callback executions (core and dependencyMgmt containers)

### PR DESCRIPTION
**Pull request by breskeby**
### Context
This is to support correct attribution of time spent executing container callbacks by the build scan plugin.

As we use a lot of container implementations I decided for the sake of easier reviews (this PR here already contains *60 changed files* ) to split things up into different PRs. 

This PR already contains 

1. introduce a feature toggle for decorating container callback actions  
2. contains infrastructure to decorate container callback actions
3. decorates container callbacks for container implementations of *core* and *dependencyManagement* subprojects. 

The ready to merge build shows that these changes come with a price. there is some regression coming with this. I wonder if we should keep this change behind a toggle for now till we general put some dedicated work on addressing these performance issues.
 
Tackling remaining DomainContainer implementations / occurances will be done in upcoming PRs

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status

### CODE REVIEW

**Comment by breskeby** [comment]
Moving back to WIP as I discovered an issue with filtered containers emitting build ops which they shouldn't

**Comment by breskeby** [comment]
handling filtered collection issue is solved.

File: subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollectionCallbackActionDecorator.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.RunnableBuildOperation;
+
+public class DefaultDomainObjectCollectionCallbackActionDecorator implements DomainObjectCollectionCallbackActionDecorator {
+    private final BuildOperationExecutor buildOperationExecutor;
+    private final UserCodeApplicationContext userCodeApplicationContext;
+
+    public DefaultDomainObjectCollectionCallbackActionDecorator(BuildOperationExecutor buildOperationExecutor, UserCodeApplicationContext userCodeApplicationContext) {
+        this.buildOperationExecutor = buildOperationExecutor;
+        // TODO RG: discuss interface
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollectionCallbackActionDecorator.java]: leftover from before changing the UserCodeApplicationContext interface

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/WrapperPlugin.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
 public class WrapperPlugin implements Plugin<Project> {
     public void apply(Project project) {
         if (project.getParent() == null) {
-            project.getTasks().register("wrapper", Wrapper.class, new Action<Wrapper>() {
+            project.getTasks().register("wrapper", Wrapper.class, new InternalAction<Wrapper>() {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/WrapperPlugin.java]: fixed

File: subprojects/core/src/main/java/org/gradle/api/internal/BuildOperationActionDecoratingCollectionEventRegistrar.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+package org.gradle.api.internal;
+
+import org.gradle.api.Action;
+import org.gradle.api.internal.collections.CollectionEventRegister;
+import org.gradle.internal.ImmutableActionSet;
+
+import javax.annotation.Nullable;
+
+public class BuildOperationActionDecoratingCollectionEventRegistrar<T> implements CollectionEventRegister<T> {
+    private CollectionEventRegister<T> delegate;
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/BuildOperationActionDecoratingCollectionEventRegistrar.java]: fixed

File: subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java [hunk 4]
```diff
@@ -80,6 +84,10 @@ protected void realized(ProviderInternal<? extends T> provider) {
         return store;
     }
 
+    public DomainObjectCollectionCallbackActionDecorator getCallbackActionDecorator() {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java]: fixed

File: subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java [hunk 5]
```diff
@@ -138,7 +146,7 @@ protected void realized(ProviderInternal<? extends T> provider) {
 
     public void all(Action<? super T> action) {
         assertMutable("all(Action)");
-        action = addEagerAction(action);
+        Action<? super T> eagerAction= addEagerAction(action);
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java]: fixed

File: subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java [hunk 6]
```diff
@@ -826,8 +826,9 @@ public String toString() {
     }
 
     protected class ExistingNamedDomainObjectProvider<I extends T> extends AbstractNamedDomainObjectProvider<I> {
+
         public ExistingNamedDomainObjectProvider(String name, Class type) {
-           super(name, type);
+            super(name, type);
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java]: fixed

File: subprojects/core/src/main/java/org/gradle/api/internal/project/BuildOperationCrossProjectConfigurator.java [hunk 7]
```diff
@@ -143,6 +143,7 @@ private CrossConfigureProjectBuildOperation(Project project) {
 
         @Override
         public BuildOperationDescriptor.Builder description() {
+
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/project/BuildOperationCrossProjectConfigurator.java]: removed

File: subprojects/core/src/test/groovy/org/gradle/api/internal/CompositeDomainObjectSetTest.groovy [hunk 8]
```diff
@@ -363,7 +363,10 @@ class CompositeDomainObjectSetTest extends Specification {
         def calledFor = []
 
         when:
-        composite.all { calledFor << it }
+        composite.all {
+            println 'calledFor ' + it
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/CompositeDomainObjectSetTest.groovy]: removed

File: subprojects/diagnostics/src/main/java/org/gradle/api/plugins/HelpTasksPlugin.java [hunk 9]
```diff
@@ -67,7 +67,7 @@ public void apply(final ProjectInternal project) {
         tasks.register(DEPENDENT_COMPONENTS_TASK, DependentComponentsReport.class, new DependentComponentsReportAction(projectName));
     }
 
-    private static class HelpAction implements Action<Help> {
+    private static class HelpAction implements InternalAction<Help> {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/diagnostics/src/main/java/org/gradle/api/plugins/HelpTasksPlugin.java]: removed

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java [hunk 10]
```diff
@@ -37,12 +37,23 @@
...
     protected AbstractNamedDomainObjectContainer(Class<T> type, Instantiator instantiator, Namer<? super T> namer) {
-        super(type, instantiator, namer);
+        this(type, instantiator, namer, DomainObjectCollectionCallbackActionDecorator.NOOP);
+    }
+
+    protected AbstractNamedDomainObjectContainer(Class<T> type, Instantiator instantiator, DomainObjectCollectionCallbackActionDecorator callbackActionDecorator) {
+        super(type, instantiator, Named.Namer.forType(type), callbackActionDecorator);
     }
 
     protected AbstractNamedDomainObjectContainer(Class<T> type, Instantiator instantiator) {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java]: yes, once all implementations have been adopted

File: subprojects/core/src/main/java/org/gradle/api/internal/BuildOperationActionDecoratingCollectionEventRegistrar.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+package org.gradle.api.internal;
+
+import org.gradle.api.Action;
+import org.gradle.api.internal.collections.CollectionEventRegister;
+import org.gradle.internal.ImmutableActionSet;
+
+import javax.annotation.Nullable;
+
+public class BuildOperationActionDecoratingCollectionEventRegistrar<T> implements CollectionEventRegister<T> {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/BuildOperationActionDecoratingCollectionEventRegistrar.java]: The problem I faced is that when moving this decoration to BroadcastingCollectionEventRegister, the order of decorating actions is done in the wrong order when wrapping the eventRegistrar with FlushingEventRegister as the filtering logic added by FlushingEventRegister is done after the build operation is fired.
Therefore the BuildOperationActionDecoratingCollectionEventRegistrar always needs to be the outer decoration layer

File: subprojects/core/src/main/java/org/gradle/api/internal/BuildOperationActionDecoratingCollectionEventRegistrar.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public Action<? super T> registerEagerAddAction(Class<? extends T> type, Action<? super T> addAction) {
+        return delegate.registerEagerAddAction(type, decorator.decorate(addAction));
+    }
+
+    @Override
+    public Action<? super T> registerLazyAddAction(Action<? super T> addAction) {
+        if (decorator == null) {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/BuildOperationActionDecoratingCollectionEventRegistrar.java]: removed.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractValidatingNamedDomainObjectContainer.java [hunk 13]
```diff
@@ -31,14 +31,25 @@
 
     private final String nameDescription;
 
+    protected AbstractValidatingNamedDomainObjectContainer(Class<T> type, Instantiator instantiator, Namer<? super T> namer, CollectionCallbackActionDecorator callbackActionDecorator) {
+        super(type, instantiator, namer, callbackActionDecorator);
+        nameDescription = type.getSimpleName() + " name";
+    }
+
+    /**
+     * TODO: remove when all implementations ported to use a CollectionCallbackActionDecorator
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractValidatingNamedDomainObjectContainer.java]: When will this be addressed?

File: subprojects/core/src/main/java/org/gradle/api/internal/collections/CollectionEventRegister.java [hunk 14]
```diff
@@ -32,9 +33,13 @@
...
+    Action<? super T> registerEagerAddAction(Class<? extends T> type, Action<? super T> addAction);
 
-    void registerLazyAddAction(Action<? super T> addAction);
+    Action<? super T> registerLazyAddAction(Action<? super T> addAction);
 
     void registerRemoveAction(Class<? extends T> type, Action<? super T> removeAction);
+
+    <S extends T> CollectionEventRegister<S> filtered(CollectionFilter<S> filter);
+
+    CollectionCallbackActionDecorator getDecorator();
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/collections/CollectionEventRegister.java]: Maybe just add a `decorate` method here instead?

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractValidatingNamedDomainObjectContainer.java (hunk 13 above)
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractValidatingNamedDomainObjectContainer.java]: in the upcoming week / weeks

**Pull request merged**